### PR TITLE
[SPIR-V] Improve type inference for a known instruction's builtin: OpGroupAsyncCopy

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR improves type inference for a known instruction's builtin: OpGroupAsyncCopy:
* deduce a type of one source/destination pointer when it's possible to deduce a type of another argument, and
* validate src and dest types and tries to unfold a parameter if it's a structure wrapper around a scalar/vector type.

### CODE REVIEW

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp [hunk 1]
```diff
@@ -2377,9 +2383,80 @@ static bool generateLoadStoreInst(const SPIRV::IncomingCall *Call,
...
 }
 
-/// Lowers a builtin funtion call using the provided \p DemangledCall skeleton
-/// and external instruction \p Set.
 namespace SPIRV {
+// Try to find a builtin funtion attributes by a demangled function name and
+// return a tuple <builtin group, op code, ext instruction number>, or a special
+// tuple value <-1, 0, 0> if the builtin funtion is not found.
+// Not all builtin funtions are supported, only those with a ready-to-use op
+// code or instruction number defined in TableGen.
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp:2391]: Three typos: funtion -> function

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp [hunk 2]
```diff
@@ -2377,9 +2383,80 @@ static bool generateLoadStoreInst(const SPIRV::IncomingCall *Call,
...
+// TODO: consider a major rework of mapping demangled calls into a builtin
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp:2392]: Just a comment, not necessary to address right now, we might be able to implement this mapping function directly in the TableGen and expose as a lookup function. Though I am not totally sure.

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.h [hunk 3]
```diff
@@ -38,6 +38,12 @@ std::optional<bool> lowerBuiltin(const StringRef DemangledCall,
                                  const SmallVectorImpl<Register> &Args,
                                  SPIRVGlobalRegistry *GR);
 
+/// Helper external function for finding a builtin funtion attributes
```
Comment (michalpaszkowski) [inline llvm/lib/Target/SPIRV/SPIRVBuiltins.h:41]: Typo: funtion -> function

Why do you say this function is external? I think all functions declared in SPIRVBuiltins.h are defined in SPIRVBuiltins.cpp

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp (hunk 2 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp:2392]: Yes, I agree that we should consider improving the whole thing. I've been thinking about the same matter when created this PR, so I've added TODO to keep that topic in plans.

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.h (hunk 3 above)
Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVBuiltins.h:41]: Thank you, I fix this

Comment (VyacheslavLevytskyy) [reply llvm/lib/Target/SPIRV/SPIRVBuiltins.h]: @michalpaszkowski done

**Review by michalpaszkowski — APPROVED** [review]

File: llvm/lib/Target/SPIRV/SPIRVBuiltins.h (hunk 3 above)
Comment (michalpaszkowski) [reply llvm/lib/Target/SPIRV/SPIRVBuiltins.h]: Thank you!

**Pull request merged**